recon/src/protocol.rs

```diff
@@ -735,10 +783,35 @@ where
     }
 
     async fn persist_all(&mut self) -> Result<()> {
+        if self.event_q.is_empty() {
+            return Ok(());
+        }
+
         let evs: Vec<_> = self.event_q.drain(..).collect();
-        if !evs.is_empty() {
-            self.recon.insert(evs).await?;
+
+        let mut batch = self.recon.insert(evs).await.context("persisting all")?;
+        if !batch.invalid.is_empty() {
+            if let Ok(cnt) = batch.invalid.len().try_into() {
+                self.recon.metrics().record(&InvalidEvents(cnt))
+            }
+            tracing::warn!(
+                "Recon discovered data it will never allow. Hanging up on peer {}",
+                self.config.peer_id
+            );
+            bail!("Received unknown data from peer: {}", self.config.peer_id);
         }
+
+        if !batch.pending.is_empty() {
+            let (added, remaining_capacity) = self.pending.track_pending(&mut batch.pending);
+            if let Ok(cnt) = added.try_into() {
+                self.recon.metrics().record(&PendingEvents(cnt))
+            }
+            if remaining_capacity < self.config.max_pending_items / 10 {
```

Comment:
what happens when capacity hits zero?

Reply:
since I default max_pending_items to a minimum of ten, it will start logging a lot because 0 < 1.. it's debug mode but probably better to not log thousands of messages just because we ran out.. i'm not sure if we want/need this message at all tbh and can just rely on something in prometheus